src/udev/udev-node.c

```diff
@@ -278,10 +280,51 @@ static int update_stack_dir(sd_device *dev, const char *dirname, bool add) {
                 return log_oom_debug();
 
         if (!add) {
-                if (unlink(filename) < 0 && errno != ENOENT)
-                        log_device_debug_errno(dev, errno, "Failed to remove %s, ignoring: %m", filename);
+                bool unlink_failed = false;
 
-                (void) rmdir(dirname);
+                if (stat(dirname, &st1) < 0) {
+                        if (errno == ENOENT)
+                                return 0;
+                        log_device_debug_errno(dev, errno, "Failed to stat %s, ignoring: %m", dirname);
+                }
+
+                if (unlink(filename) < 0) {
+                        unlink_failed = true;
+                        if (errno != ENOENT)
+                                log_device_debug_errno(dev, errno, "Failed to remove %s, ignoring: %m", filename);
+                }
+
+                if (rmdir(dirname) >= 0 || errno == ENOENT)
+                        return 0;
+
+                if (unlink_failed)
+                        return 0;
+
+                if ((st1.st_mode & S_IFMT) == 0)
+                        return 0;
+
+                /* symlink was removed. Check if the timestamp of directory is changed. */
+                for (unsigned k = 0; k < UPDATE_TIMESTAMP_MAX_RETRIES; k++) {
+                        if (stat(dirname, &st2) < 0) {
+                                if (errno != ENOENT)
+                                        log_device_debug_errno(dev, errno, "Failed to stat %s, ignoring: %m", dirname);
+                                return 0;
+                        }
+
+                        if (!stat_inode_unmodified(&st1, &st2))
+                                return 0;
+
+                        log_device_debug(dev, "Symlink %s is removed, but timestamp of %s is not changed, updating timestamp after 10ms.",
+                                         filename, dirname);
+                        (void) usleep(10 * USEC_PER_MSEC);
+                        if (utimensat(AT_FDCWD, dirname, NULL, 0) < 0) {
```

Comment:
hmm, why is this necssary? we operate on tmpfs here, we know that mtimes are properly updated there, why do this explicitly here?

Reply:
Please read the comments in the revised code.